feat: add prefer-find-by rule

gndelia opened the pull request:
This one tackle #127 

While current code works, I still have to polish some stuff and add more scenarios (and update the docs)

Belco90 commented on docs/rules/prefer-find-by.md:
```diff
@@ -0,0 +1,43 @@
+
+## Further Reading
+
+TBD
```
I would link to [findBy query doc](https://testing-library.com/docs/dom-testing-library/api-queries#findby) here, as it mentions that `findBy` is a shortcut for `waitFor` + `getBy`. I would use the text from that callout as the main inspiration for this url in the description at the top.

Belco90 commented:
Looking good so far!

benmonro commented on docs/rules/prefer-find-by.md:
````diff
@@ -0,0 +1,43 @@
+```js
+const submitButton = await findByText('foo');
+
+const submitButton = await screen.findAllByRole('table');
````
I would add some correct usages of waitFor here as well.

benmonro commented on docs/rules/prefer-find-by.md:
```diff
@@ -0,0 +1,43 @@
+
+## When Not To Use It
+
+TBD
```
```suggestion
Don't use those rule if you don't care if people aren't following best practices of testing library.
```

Belco90 replied:
I wouldn't go that _direct_ here. Better something nicer like "Not encouraging use of `findBy` shortcut from testing library best practices".

benmonro commented on lib/rules/prefer-find-by.ts:
```diff
@@ -0,0 +1,88 @@
+    docs: {
+      description: 'Suggest using find* instead of waitFor to wait for elements',
+      category: 'Best Practices',
+      recommended: false,
```
Why not recommend this one?

Belco90 replied:
I think it should be recommended, indeed.

gndelia replied:
"warn" or "error" ?

Belco90 replied:
Nice question. I think warn is enough? This rules doesn't prevent any malfunctioning code, just warns about a preferred way of writing it.

gndelia replied:
I added `warn`

benmonro commented on lib/rules/prefer-find-by.ts:
```diff
@@ -0,0 +1,88 @@
+    messages: {
+      preferFindBy: 'Prefer {{queryVariant}}{{queryMethod}} method over using await {{fullQuery}}'
+    },
+    fixable: null,
```
Can we not make the simple ones fixable?

Belco90 replied:
This is still a draft Ben. Leave the guy work on this! We are discussing in the original issue if addressing it here or in a 2nd PR.

benmonro replied:
Was just an idea.  Not trying to clock this or anything...

gndelia commented:
I pushed a new version.

- CI passes: More test cases!
- Docs updated: I added links, details, and more examples

still pending to make it fixable

Belco90 commented on lib/rules/prefer-find-by.ts:
```diff
@@ -0,0 +1,88 @@
+
+type Options = [];
+export type MessageIds = 'preferFindBy';
+// TODO check if this should be under utils.ts - there are some async utils
```
We have `ASYNC_UTILS` in `utils.ts` but it includes more methods than here.

Belco90 replied:
Maybe it's a good idea to keep all different async utils under an enumerate type and then build `ASYNC_UTILS` all the enumerate options?

gndelia replied:
after giving some though, I am not sure of moving this one out. It's not a list of all `wait` methods, but just a list of the ones this rule cares about. Not sure how useful it could be outside of here, as it's not _all_ the `wait` methods. Either way, we could centralize them in the future if we see the chance

Belco90 commented:
> I pushed a new version.
> 
>     * CI passes: More test cases!
> 
>     * Docs updated: I added links, details, and more examples
> 
> 
> still pending to make it fixable

Nice work mate. It would be nice to have the fixable code here, but let me know what you think about it and if you prefer to address it in a separated PR in the future.

gndelia commented:
This PR is ready to be reviewed.

I gave [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) a try (instead of directly fixing the code), but the problem is that types in ESLint 6 differ from ESLint 7. I have a local branch with the `suggestions` which _I believe_ are working, but the tests always pass for the suggestions; It's like I cannot force them to fail. I believe that once #138  is finished, suggestions should work better

So this PR goes without suggestions for fix available 😞

Belco90 commented on docs/rules/prefer-find-by.md:
```diff
@@ -0,0 +1,74 @@
+  screen.getByRole('button', { name: /submit/i })
+);
+const submitButton = await waitFor(() =>
+  screen.getAllTestId('button', { name: /submit/i })
```
```suggestion
  screen.getAllByTestId('button', { name: /submit/i })
```

Belco90 commented on docs/rules/prefer-find-by.md:
```diff
@@ -0,0 +1,74 @@
+});
+
+// using a custom arrow function
+await waitFor(() => myCustomFunction());
```
Can you add an extra correct usage of `waitFor` within expects?

```javascript
await waitFor(() => expect(screen.getByText('bar').toBeDisabled());
```

gndelia replied:
done

Belco90 commented on tests/lib/rules/prefer-find-by.test.ts:
```diff
@@ -0,0 +1,80 @@
+          return ${queryMethod}()
+        })
+      `
+    })),
```
Could you add valid tests for the example requested in the doc? A valid `waitFor` returning expect with getBy inside.

gndelia replied:
done

gndelia commented:
BTW, I added the new rule to the main README.

![image](https://user-images.githubusercontent.com/352474/83360577-aa6de600-a358-11ea-9282-743b205c8732.png)




> Looking good mate, thanks for your work! I just requested couple of extra examples and tests, everything else seems fine. We are leaving the fix step for a future PR then right?

yes. The problem is that when testing `suggestions`, no matter what the `expected` output I write, the tests always passes. Let's look at this example.

```ts
code: `
            const submitButton = await ${waitMethod}(() => screen.${queryMethod}('foo', { name: 'baz' }))
          `,
          errors: [{
            messageId: 'preferFindBy',
            data: {
              queryVariant: queryMethod.includes('All') ? 'findAllBy': 'findBy',
              queryMethod: queryMethod.split('By')[1],
              fullQuery: `${waitMethod}(() => screen.${queryMethod}('foo', { name: 'baz' }))`,
            },
            suggestions: [{
              messageId: 'preferFindBy',
              // this expected output should match the suggestion output of the rule
              output: `const submitButton = await screen.${queryMethod.includes('All') ? 'findAllBy': 'findBy'}${queryMethod.split('By')[1]}('foo', { name: 'baz' })`
            }]
          }]
        })
```

No matter what output I write in the test, the tests pass, making me think that there's something wrong with the test assertion declaration.

I think the problem might be solved with ESlint 7 - furthermore, the contract changed in that version so it'll be easier to implement once that is solved.

We could probably go with that in another PR, does that sound good for you?

Belco90 commented:
@gndelia that sounds good. I remember I had a similar problem when implementing the fix for the `waitFor` migration. We can wait a little bit until we find proper way to test it, but I think we should go for fix rather than suggestion as there is just single way of changing a getBy + waitFor by a findBy.

Belco90 commented on docs/rules/prefer-find-by.md:
```diff
@@ -0,0 +1,78 @@
+# Use `find*` query methods to wait for elements instead of waitFor (prefer-find-by)
```
Can you use here same description from README please?

```suggestion
# Suggest using `findBy*` methods instead of the `waitFor` + `getBy` queries (prefer-find-by)
```

gndelia replied:
done

Belco90 commented:
I've been thinking about this rule name. Could we regret it in the future for being so generic? Could we have in the future other cases where we prefer `findBy` over other options? I don't know if could be worth it to think about a better name for the rule. ESLint rules names should be short tho, but I'd like to think about it. Could something like `no-get-by-wait-for-combo` work? Maybe I'm overthinking around this.

benmonro commented:
I think when it comes to rule names less is more. Either prefer-find-by or no-wait-for-get

gndelia commented:
we could expand the rules or rename (breaking change) if we get a collision in the future. Not sure which other case `findBy*` offers, to prefer it over something else...

Belco90 commented:
:tada: This PR is included in version 3.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/eslint-plugin-testing-library)
- [GitHub release](https://github.com/testing-library/eslint-plugin-testing-library/releases/tag/v3.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
